File: day10/python/subesokun/main.py
INPUT_FILE_NAME = 'input.txt'

def parseSpotInput(text):
    tmp_split = text.split('>')
    pos_str, velo_str = tmp_split[0], tmp_split[1]
    pos = tuple([int(coord) for coord in pos_str.replace(' ', '').split('<')[1].split(',')])
    velo = tuple([int(velo) for velo in velo_str.replace(' ', '').split('<')[1].split(',')])
    return [pos, velo]

def hasSpotAt(x, y, spots):
    for spot in spots:
        if spot[0] == (x,y):
            return True
    return False

def printSpots(spots):
    min_x = min(spots, key = lambda s: s[0][0])[0][0]
    max_x = max(spots, key = lambda s: s[0][0])[0][0]
    min_y = min(spots, key = lambda s: s[0][1])[0][1]
    max_y = max(spots, key = lambda s: s[0][1])[0][1]
    for y in range(min_y, max_y + 1):
        print(''.join(['#' if hasSpotAt(x, y, spots) else ' ' for x in range(min_x, max_x + 1)]))

def tick(spots, dir=1):
    for spot in spots:
        spot[0] = (spot[0][0] + dir * spot[1][0], spot[0][1] + dir * spot[1][1])

def searchAndPrint(spots):
    current_min_delta_y = None
    current_second = 0
    while True:
        tick(spots)
        current_second += 1
        min_y = min(spots, key = lambda s: s[0][1])[0][1]
        max_y = max(spots, key = lambda s: s[0][1])[0][1]
        min_delta_y = max_y - min_y
        if current_min_delta_y == None or min_delta_y < current_min_delta_y:
            current_min_delta_y = min_delta_y
        else:
            tick(spots, -1)
            printSpots(spots)
            return current_second - 1

spots = []
with open(INPUT_FILE_NAME) as input_file:
    for line in input_file:
        spots.append(parseSpotInput(line.rstrip()))

print('Solution to part 1:')

seconds = searchAndPrint(spots)

print('Solution to part 2: %i' % (seconds,))

File: day10/go/subesokun/src/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"reflect"
	"regexp"
	"strconv"
	"strings"
)

type spot struct {
	px, py, vx, vy int
}

func (s *spot) move(dir int) {
	s.px = s.px + dir*s.vx
	s.py = s.py + dir*s.vy
}

func getField(s *spot, field string) int {
	r := reflect.ValueOf(s)
	f := reflect.Indirect(r).FieldByName(field)
	return int(f.Int())
}

func parseSpotInput(text string) spot {
	r := regexp.MustCompile(`position=<\s*(?P<px>-?\d+),\s*(?P<py>-?\d+)>\svelocity=<\s*(?P<vx>-?\d+),\s*(?P<vy>-?\d+)`)
	matches := r.FindStringSubmatch(text)
	px, _ := strconv.Atoi(matches[1])
	py, _ := strconv.Atoi(matches[2])
	vx, _ := strconv.Atoi(matches[3])
	vy, _ := strconv.Atoi(matches[4])
	return spot{px: px, py: py, vx: vx, vy: vy}
}

func minMaxSlice(spots []spot, field string) (int, int) {
	var min, max int
	for i := range spots {
		value := getField(&spots[i], field)
		if i == 0 || value < min {
			min = value
		}
		if i == 0 || value > max {
			max = value
		}
	}
	return min, max
}

func hasSpotAt(x int, y int, spots []spot) bool {
	for i := range spots {
		if spots[i].px == x && spots[i].py == y {
			return true
		}
	}
	return false
}

func tick(spots []spot, dir int) {
	for i := range spots {
		spots[i].move(dir)
	}
}

func printSpots(spots []spot) {
	minX, maxX := minMaxSlice(spots, "px")
	minY, maxY := minMaxSlice(spots, "py")
	var plot []string
	for y := minY; y <= maxY; y++ {
		for x := minX; x <= maxX; x++ {
			if hasSpotAt(x, y, spots) {
				plot = append(plot, "#")
			} else {
				plot = append(plot, " ")
			}
		}
		plot = append(plot, "\n")
	}
	fmt.Println(strings.Join(plot, ""))
}

func searchAndPrint(spots []spot) int {
	var deltaY, minY, maxY int
	currentMinDeltaY := 0
	currentSecond := 0
	for true {
		tick(spots, 1)
		currentSecond++
		minY, maxY = minMaxSlice(spots, "py")
		deltaY = maxY - minY
		if currentMinDeltaY == 0 || deltaY < currentMinDeltaY {
			currentMinDeltaY = deltaY
		} else {
			tick(spots, -1)
			printSpots(spots)
			break
		}
	}
	return currentSecond - 1
}

func main() {
	var spots []spot
	file, _ := os.Open("input.txt")
	defer file.Close()
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		spots = append(spots, parseSpotInput(scanner.Text()))
	}

	fmt.Printf("Solution to part 1:\n")
	seconds := searchAndPrint(spots)
	fmt.Printf("Solution to part 2: %v\n", seconds)
}
